File: examples/cpp/run_async_model_profiler/run_async_model_profiler.cpp
/*
 * Copyright (C) 2018- DEEPX Ltd.
 * All rights reserved.
 *
 * This software is the property of DEEPX and is provided exclusively to customers
 * who are supplied with DEEPX NPU (Neural Processing Unit).
 * Unauthorized sharing or usage is strictly prohibited by law.
 */

/**
 * @file run_async_model_profiler.cpp
 * @brief Async inference with per-job profiling using the Wait pattern.
 *
 * Follows the run_async_model_wait flow:
 *   Main thread  : RunAsync() × N  →  push jobId to queue
 *   Worker thread: pop jobId → ie.Wait(jobId) → profiler.GetJobMetrics(jobId) → print
 *
 * This gives accurate per-job H2D / NPU / D2H / Task timings for each job.
 *
 * Usage:
 *   run_async_model_profiler -m <model.dxnn> [-l <loops>] [--use-ort]
 */

#include "dxrt/dxrt_cxx_api.h"
#include "dxrt/extern/cxxopts.hpp"
#include "../include/logger.h"
#include "../include/concurrent_queue.h"

#include <iomanip>
#include <iostream>
#include <string>
#include <vector>
#include <thread>

static ConcurrentQueue<int> gJobIdQueue(32);

static void printJobMetrics(dxrt::Logger& log, int jobIdx, int jobId,
                            const dxrt::JobMetrics& jm)
{
    if (!jm.valid)
    {
        log.Info("[Job #" + std::to_string(jobIdx) + " id=" + std::to_string(jobId) +
                 "] No profiler data");
        return;
    }

    for (const auto& task : jm.tasks)
    {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(3);
        oss << "[Job #" << jobIdx << " id=" << jobId << "] task=" << task.task_name;

        for (const auto& devPair : task.devices)
        {
            const int devId  = devPair.first;
            const auto& d    = devPair.second;
            oss << "  |  Dev" << devId
                << ": InputNFH=" << d.input_format_us << "us"
                << " H2D=" << d.h2d_us  << "us"
                << " Inference=" << d.inference_core_all_us << "us"
                << " (Core0=" << d.inference_core_0_us
                << " Core1=" << d.inference_core_1_us
                << " Core2=" << d.inference_core_2_us << ")"
                << " D2H=" << d.d2h_us  << "us"
                << " OutputNFH=" << d.output_format_us << "us"
                << " NPU Task=" << d.total_us << "us";
        }

        if (task.cpu_task_us > 0.0)
        {
            oss << "  |  CPU Task=" << task.cpu_task_us << "us";
        }

        log.Info(oss.str());
    }
}

// Worker thread: Wait for each job and print profiling metrics
static int waitThreadFunc(const dxrt::InferenceEngine& ie,
                          dxrt::Profiler& profiler,
                          int loopCount)
{
    static const auto& log = dxrt::Logger::GetInstance();

    for (int idx = 0; idx < loopCount; ++idx)
    {
        int jobId = gJobIdQueue.pop();

        try
        {
            auto outputs = ie.Wait(jobId);
            (void)outputs;
        }
        catch (const dxrt::Exception& e)
        {
            log.Error(std::string(e.what()) +
                      " error-code=" + std::to_string(static_cast<int>(e.code())));
            return -1;
        }

        auto metrics = profiler.GetJobMetrics(jobId);
        printJobMetrics(const_cast<dxrt::Logger&>(log), idx, jobId, metrics);
    }

    return 0;
}

int main(int argc, char* argv[])
{
    std::string model_path;
    int         loop_count;
    bool        use_ort;
    bool        verbose;

    auto& log = dxrt::Logger::GetInstance();

    cxxopts::Options options("run_async_model_profiler",
                             "Async inference with per-job profiling (Wait + GetJobMetrics)");
    options.add_options()
        ("m,model",   "Path to model file (.dxnn)",
         cxxopts::value<std::string>(model_path))
        ("l,loops",   "Number of inference loops",
         cxxopts::value<int>(loop_count)->default_value("10"))
        ("use-ort",   "Enable ORT (CPU post-processing) for the model",
         cxxopts::value<bool>(use_ort)->default_value("false"))
        ("v,verbose", "Enable verbose/debug logging",
         cxxopts::value<bool>(verbose)->default_value("false"))
        ("h,help",    "Print usage");

    try
    {
        auto result = options.parse(argc, argv);
        if (result.count("help") || !result.count("model"))
        {
            std::cout << options.help() << std::endl;
            return result.count("help") ? 0 : -1;
        }
        if (verbose)
        {
            log.SetLevel(dxrt::Logger::Level::LOGLEVEL_DEBUG);
        }
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("Error parsing arguments: ") + e.what());
        std::cout << options.help() << std::endl;
        return -1;
    }

    log.Info("Model  : " + model_path);
    log.Info("Loops  : " + std::to_string(loop_count));
    log.Info("UseORT : " + std::string(use_ort ? "true" : "false"));

    try
    {
        // ── Enable profiler ───────────────────────────────────────────────
        auto& config = dxrt::Configuration::GetInstance();
        config.SetEnable(dxrt::Configuration::ITEM::PROFILER, true);
        config.SetAttribute(dxrt::Configuration::ITEM::PROFILER,
                            dxrt::Configuration::ATTRIBUTE::PROFILER_SHOW_DATA, "ON");
        config.SetAttribute(dxrt::Configuration::ITEM::PROFILER,
                            dxrt::Configuration::ATTRIBUTE::PROFILER_SAVE_DATA, "OFF");

        auto& profiler = dxrt::Profiler::GetInstance();

        // ── Create InferenceEngine ────────────────────────────────────────
        dxrt::InferenceOption option;
        option.useORT = use_ort;
        dxrt::InferenceEngine ie(model_path, option);

        // ── Start worker thread ───────────────────────────────────────────
        auto worker = std::thread(waitThreadFunc, std::ref(ie), std::ref(profiler), loop_count);

        // ── Submit all jobs ───────────────────────────────────────────────
        std::vector<uint8_t> input(ie.GetInputSize(), 0);

        auto start = std::chrono::high_resolution_clock::now();

        for (int i = 0; i < loop_count; ++i)
        {
            int jobId = ie.RunAsync(input.data());
            gJobIdQueue.push(jobId);
            log.Debug("Submitted jobId=" + std::to_string(jobId));
        }

        worker.join();

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> elapsed = end - start;

        double total_ms = elapsed.count();
        double avg_ms   = total_ms / static_cast<double>(loop_count);
        double fps      = 1000.0 / avg_ms;

        log.Info("-----------------------------------");
        log.Info("Total   : " + std::to_string(total_ms) + " ms");
        log.Info("Average : " + std::to_string(avg_ms)   + " ms/job");
        log.Info("FPS     : " + std::to_string(fps));
        log.Info("-----------------------------------");


    }
    catch (const dxrt::Exception& e)
    {
        log.Error(std::string("dxrt::Exception: ") + e.what());
        return -1;
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("std::exception: ") + e.what());
        return -1;
    }
    catch (...)
    {
        log.Error("Unknown exception");
        return -1;
    }

    return 0;
}

File: examples/python/run_async_model_profiler.py
#
# Copyright (C) 2018- DEEPX Ltd.
# All rights reserved.
#
# This software is the property of DEEPX and is provided exclusively to customers
# who are supplied with DEEPX NPU (Neural Processing Unit).
# Unauthorized sharing or usage is strictly prohibited by law.
#

"""Async inference with per-job profiling using the Wait pattern.

Follows the run_async_model_wait flow:
  Main thread  : run_async() × N  →  push job_id to queue
  Worker thread: pop job_id → ie.wait(job_id) → profiler.get_job_metrics(job_id) → print

This gives accurate per-job H2D / NPU / D2H / Task timings for each job.

Usage:
    python run_async_model_profiler.py -m <model.dxnn> [-l <loops>] [--use-ort]
"""

import argparse
import os
import queue
import threading
import time

import numpy as np
from dx_engine import Configuration, InferenceEngine, InferenceOption, Profiler

# ---------------------------------------------------------------------------
# Job-ID queue shared between main thread and worker thread
# ---------------------------------------------------------------------------
g_job_id_queue: queue.Queue = queue.Queue(maxsize=32)


# ---------------------------------------------------------------------------
# Per-job metric printing
# ---------------------------------------------------------------------------

def print_job_metrics(job_idx: int, job_id: int, jm) -> None:
    if not jm.valid:
        print(f"[Job #{job_idx} id={job_id}] No profiler data")
        return

    for task in jm.tasks:
        parts = [f"[Job #{job_idx} id={job_id}] task={task.task_name}"]

        for dev_id, d in task.devices.items():
            parts.append(
                f"  |  Dev{dev_id}:"
                f" InputNFH={d.input_format_us:.3f}us"
                f" H2D={d.h2d_us:.3f}us"
                f" Inference={d.inference_core_all_us:.3f}us"
                f" (Core0={d.inference_core_0_us:.3f}"
                f" Core1={d.inference_core_1_us:.3f}"
                f" Core2={d.inference_core_2_us:.3f})"
                f" D2H={d.d2h_us:.3f}us"
                f" OutputNFH={d.output_format_us:.3f}us"
                f" NPU Task={d.total_us:.3f}us"
            )

        if task.cpu_task_us > 0.0:
            parts.append(f"  |  CPU Task={task.cpu_task_us:.3f}us")

        print("".join(parts))


# ---------------------------------------------------------------------------
# Worker thread: Wait for each job and print profiling metrics
# ---------------------------------------------------------------------------

def wait_thread_func(ie: InferenceEngine, profiler: Profiler, loop_count: int) -> None:
    for idx in range(loop_count):
        job_id = g_job_id_queue.get()

        try:
            outputs = ie.wait(job_id)
            _ = outputs
        except Exception as e:
            print(f"[Worker] Exception: {e}")
            return

        jm = profiler.get_job_metrics(job_id)
        print_job_metrics(idx, job_id, jm)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Async inference with per-job profiling (Wait + get_job_metrics)"
    )
    parser.add_argument("--model", "-m", type=str, required=True,
                        help="Path to model file (.dxnn)")
    parser.add_argument("--loops", "-l", type=int, default=10,
                        help="Number of inference loops (default: 10)")
    parser.add_argument("--use-ort", action="store_true",
                        help="Enable ORT (CPU post-processing) for the model")
    parser.add_argument("--verbose", "-v", action="store_true",
                        help="Enable verbose output")
    args = parser.parse_args()

    if not os.path.exists(args.model):
        parser.error(f"Model path '{args.model}' does not exist.")

    return args


if __name__ == "__main__":
    args = parse_args()

    print(f"Model  : {args.model}")
    print(f"Loops  : {args.loops}")
    print(f"UseORT : {args.use_ort}")

    # ── Enable profiler ───────────────────────────────────────────────────
    config = Configuration()
    config.set_enable(Configuration.ITEM.PROFILER, True)
    config.set_attribute(Configuration.ITEM.PROFILER,
                         Configuration.ATTRIBUTE.PROFILER_SHOW_DATA, "ON")
    config.set_attribute(Configuration.ITEM.PROFILER,
                         Configuration.ATTRIBUTE.PROFILER_SAVE_DATA, "OFF")

    profiler = Profiler.get_instance()

    # ── Create InferenceEngine ────────────────────────────────────────────
    opt = InferenceOption()
    opt.use_ort = args.use_ort
    with InferenceEngine(args.model, opt) as ie:

        # ── Start worker thread ───────────────────────────────────────────
        worker = threading.Thread(
            target=wait_thread_func,
            args=(ie, profiler, args.loops),
            daemon=True,
        )
        worker.start()

        # ── Submit all jobs ───────────────────────────────────────────────
        # NOTE: np.zeros() uses COW zero pages — all virtual pages share one
        # physical page. PCIe DMA driver's get_user_pages() then sees duplicate
        # physical pages in the SG list and fails with EFAULT.
        # np.empty() + explicit fill forces unique physical page allocation.
        buf = np.empty(ie.get_input_size(), dtype=np.uint8)
        buf.fill(0)

        start = time.perf_counter()

        for i in range(args.loops):
            job_id = ie.run_async([buf])
            g_job_id_queue.put(job_id)
            if args.verbose:
                print(f"Submitted job_id={job_id}")

        worker.join()

        end = time.perf_counter()

    total_ms = (end - start) * 1000.0
    avg_ms   = total_ms / args.loops
    fps      = 1000.0 / avg_ms

    print("-----------------------------------")
    print(f"Total   : {total_ms:.2f} ms")
    print(f"Average : {avg_ms:.2f} ms/job")
    print(f"FPS     : {fps:.2f}")
    print("-----------------------------------")



